### PR TITLE
Added new low memory parser (v2)

### PR DESCRIPTION
**Pull request by ziadoz**
Based on feedback on the low memory parser v1, I've done some refactoring so our parser class can now `yield` its results, or return them in an array. The parser is now a simple class that you plug node parsers into, which makes it more flexible.

Here's a summary of what's changed: 

- Calling `parse()` uses `yield` to return each element, which results in reduced memory usage.
- Validates XML using `XMLReader` so document does not have to be loaded into memory.
- All node parsing has been abstracted to separate classes.
- We can now write and plugin our own custom node parsers for specific scenarios.
- Validation is done directly via a method, rather than in the constructor like Parser, for flexibility.

Some quick examples:

Yielding assignments: 
```
$parser = new Parser(__DIR__ . '/fixtures/assignments.xml');
$parser->validate();

foreach ($parser->parse(AssignmentNodeParser::class) as $productId => $assignment) {
    //...
}
```

Parsing into an array:

```
$parser = new Parser(__DIR__ . '/fixtures/products.xml');
$parser->validate();

$results = $parser->toArray([
    'assignments' => AssignmentNodeParser::class,
    'categories' => CategorySimpleNodeParser::class,
], ['assignments']);

print_r($results['assignments']);
print_r($results['categories']);
```

### CODE REVIEW

File: src/Parser.php [hunk 1]
```diff
@@ -1,287 +1,136 @@
...
 
-    private function addSet(string $id, SimpleXMLElement $element)
+    protected function validateNodeParserClass(string $class)
     {
-        $details = $this->commonDetails($element);
-
-        foreach ($element->{'product-set-products'}->{'product-set-product'} as $product) {
-            $details['products'][] = (string) $product['product-id'];
+        if (! is_subclass_of($class, NodeParserInterface::class)) {
+            throw new XmlException('Node parser class "' . $class . '" must implement ' . NodeParserInterface::class);
```
Comment (WillJW) [inline src/Parser.php]: Is `XmlException` correct here?

File: src/Parser/AssignmentNodeParser.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    public function isMatch(): bool
+    {
+        return ($this->reader->nodeType === XMLReader::ELEMENT && $this->reader->localName === 'category-assignment');
+    }
+
+    public function parse(): array
+    {
+        $element = new SimpleXMLElement($this->reader->readOuterXml());
+
+        $categoryId = (string) $element['category-id'];
```
Comment (WillJW) [inline src/Parser/AssignmentNodeParser.php]: Will this always be a string anyway, without needing to cast it?

File: src/Parser/BundleNodeParser.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        return $this->commonDetails($element);
+    }
+
+    public function parse(): array
+    {
+        $details = $this->getCommonDetails($this->element);
+
+        foreach ($this->element->{'bundled-products'}->{'bundled-product'} as $variation) {
+            $quantity = (int) ($variation->{'quantity'} ?? 0);
```
Comment (WillJW) [inline src/Parser/BundleNodeParser.php]: Should we group this as a one-liner?

File: src/Parser/CategoryNodeParser.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        return $this->commonDetails($element);
+    }
+
+    public function parse(): array
+    {
+        $element = new SimpleXMLElement($this->reader->readOuterXml());
+
+        return [
+            'id'   => (string)$element['category-id'],
```
Comment (WillJW) [inline src/Parser/CategoryNodeParser.php]: Should be a space in `(string) $element`

File: src/Parser/CommonDetailsNodeParserTrait.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            'parent'                  => 'parent',
+            'tax'                     => 'tax-class-id',
+            'brand'                   => 'brand',
+            'sitemap-included-flag'   => 'sitemap-included-flag',
+            'sitemap-changefrequency' => 'sitemap-changefrequency',
+            'sitemap-priority'        => 'sitemap-priority',
+        ];
+
+        foreach ($map as $name => $source) {
+            $cleansed = html_entity_decode(trim((string) $element->{$source}));
```
Comment (WillJW) [inline src/Parser/CommonDetailsNodeParserTrait.php]: We shouldn't need the curly braces around `$source`

File: src/Parser/CommonDetailsNodeParserTrait.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    protected function customAttributes(SimpleXMLElement $element): array
+    {
+        if (! isset($element->{'custom-attributes'}->{'custom-attribute'})) {
+            return [];
+        }
+
+        $attributes = [];
+
+        foreach ($element->{'custom-attributes'}->{'custom-attribute'} as $attribute) {
+            if (isset($attribute->{'value'})) {
```
Comment (WillJW) [inline src/Parser/CommonDetailsNodeParserTrait.php]: Would we simplify to `$attribute->value` here?

File: src/Parser/SetNodeParser.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    public function parse(): array
+    {
+        $details = $this->getCommonDetails($this->element);
+
+        foreach ($this->element->{'product-set-products'}->{'product-set-product'} as $product) {
+            $details['products'][] = (string) $product['product-id'];
+        }
+
+        return [
+            'id'   => (string)$this->element['product-id'],
```
Comment (WillJW) [inline src/Parser/SetNodeParser.php]: Space missing in `(string) $this`

File: src/Parser/SetSimpleNodeParser.php [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<?php
+namespace DemandwareXml\Parser;
+
+use SimpleXMLElement;
+
+class SetSimpleNodeParser extends SetNodeParser implements NodeParserInterface
+{
+    protected function getCommonDetails(SimpleXMLElement $element)
+    {
+        return $this->commonDetails($element, $skipAttributes = true);
```
Comment (WillJW) [inline src/Parser/SetSimpleNodeParser.php]: I'd remove the inline assignment here

File: src/Parser/VariationNodeParser.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    protected function getCommonDetails(SimpleXMLElement $element)
+    {
+        return $this->commonDetails($element);
+    }
+
+    public function parse(): array
+    {
+        return [
+            'id'   => (string)$this->element['product-id'],
```
Comment (WillJW) [inline src/Parser/VariationNodeParser.php]: Space missing here as well

File: tests/ParserTest.php [hunk 10]
```diff
@@ -2,122 +2,220 @@
...
 // note: don't need to parse variants, so no test for those!
 // rebuild fixtures: file_put_contents(__DIR__ . '/fixtures/categories.json', json_encode($parser->categories(), JSON_PRETTY_PRINT) . PHP_EOL);
 class ParserTest extends TestCase
 {
     use FixtureHelper;
 
-    public function testMixedParser()
+    /**
+     * @expectedException              \DemandwareXml\XmlException
+     * @expectedExceptionMessageRegExp /Fatal: xmlParseEntityRef: no name in invalid-products.xml/
```
Comment (WillJW) [inline tests/ParserTest.php]: Is this the entire error message? If so, we can do `@expectedExceptionMessage Fatal: xmlParseEntityRef: no name in invalid-products.xml`

File: src/Parser.php (hunk 1 above)
Comment (ziadoz) [reply src/Parser.php]: I've changed the exception thrown.

File: src/Parser/CategoryNodeParser.php (hunk 4 above)
Comment (ziadoz) [reply src/Parser/CategoryNodeParser.php]: Fixed.

File: tests/ParserTest.php (hunk 10 above)
Comment (ziadoz) [reply tests/ParserTest.php]: No, it's not the entire message. There's some info after that seems to differ slighty by platform, hence the regexp.

**Comment by ziadoz** [comment]
I've cleaned up some bits and pieces here.

**Comment by maxakropolis** [comment]
👍

**Comment by weshooper** [comment]
👍

**Pull request merged**